rust/build.rs

```diff
@@ -0,0 +1,46 @@
+use std::path::Path;
+
+use schemars::schema::Schema;
+use typify::{TypeSpace, TypeSpaceSettings};
+
+fn generate_schema(schema_path: &str, output_module: &str) -> String {
+    println!("cargo:rerun-if-changed={schema_path}");
+    let json = std::fs::read_to_string(schema_path).expect("Read schema JSON file");
+
+    let schema = serde_json::from_str::<schemars::schema::RootSchema>(&json).unwrap();
+
+    let mut type_space = TypeSpace::new(TypeSpaceSettings::default().with_struct_builder(true));
+
+    // TODO: when typify 0.12 is released, replace this entire block with add_root_schema
+    // see https://github.com/oxidecomputer/typify/pull/236/files
+    {
+        type_space.add_ref_types(schema.definitions).unwrap();
+
+        type_space.add_type(&Schema::Object(schema.schema)).unwrap();
+    }
+
+    let code = prettyplease::unparse(&syn::parse2::<syn::File>(type_space.to_stream()).unwrap());
+
+    format!(
+        "pub mod {output_module} {{
+    use serde::{{Deserialize, Serialize}};
+    {code}
+    }}"
+    )
+}
+
+fn main() {
+    let mut module_code = String::new();
+    module_code.push_str(&generate_schema(
+        "schemas/ingest-metrics.v1.schema.json",
+        "ingest_metrics_v1",
```

Comment:
i am not sure, right now i feel like it's not worth putting that much effort into the rust library to make that part happen

one annoying thing i ran into is that i would have to re-build topics parsing here because unlike in `generate_python_types.py` i actually cannot import the library before it is finished building

Reply:
Ok